tools/stack_decode.py

```diff
@@ -58,18 +66,32 @@ def decode_stacktrace_log(input_source):
     return
 
 
-# Output one stacktrace after passing it through addr2line with appropriate
-# options
-def output_stacktrace(thread_id, traceinfo):
-  piped_input = ""
-  for stack_addr in traceinfo.address_list:
-    piped_input += (stack_addr + "\n")
+# Execute addr2line with a particular object file and input string of addresses
+# to resolve, one per line.
+#
+# Returns list of result lines
+def run_addr2line(obj_file, piped_input):
   addr2line = subprocess.Popen(
-      ["addr2line", "-Cpisfe", traceinfo.obj_file],
+      ["addr2line", "-Cpisfe", obj_file],
       stdin=subprocess.PIPE,
       stdout=subprocess.PIPE)
   output_stdout, _ = addr2line.communicate(piped_input)
-  output_lines = output_stdout.split("\n")
+  return output_stdout.split("\n")
+
+
+# Output one stacktrace after passing it through addr2line with appropriate
+# options
+def output_stacktrace(thread_id, traceinfo):
+  output_lines = []
+  for address_list in traceinfo.obj_list:
+    piped_input = ""
+    obj_name = address_list.obj_file
+    # end of stack sentinel
+    if obj_name == "[0xffffffffffffffff]":
```

Comment:
If I make the other change below to stop before the last item then this can be removed too.